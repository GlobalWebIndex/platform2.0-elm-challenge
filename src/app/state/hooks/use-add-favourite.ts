import { useMutation, useQueryClient } from '@tanstack/react-query';
import { queryKeyFavourite } from './use-favourite.tsx';
import { addToFavourite } from '../../repository/favourite-helper.ts';

export const useAddFavourite = () => {
  const queryClient = useQueryClient();

  const { mutate: addFavourite, isPending } = useMutation({
    mutationFn: addToFavourite,
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: [queryKeyFavourite] });
    },
  });

  return {
    addFavourite,
    isPending,
  };
};
