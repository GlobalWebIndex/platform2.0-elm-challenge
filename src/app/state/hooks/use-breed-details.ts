import { useQuery } from '@tanstack/react-query';
import { fetchBreedDetails } from '../../repository/fetch-breed-details.ts';

export const queryKeyBreedDetails = 'breed-details';
export const useBreedDetails = (id: string) => {
  const { isPending, error, data } = useQuery({
    queryKey: [queryKeyBreedDetails, id],
    queryFn: () => fetchBreedDetails(id),
  });

  return {
    data,
    error,
    isPending,
  };
};
