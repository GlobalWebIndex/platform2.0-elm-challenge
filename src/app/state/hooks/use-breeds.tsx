import { useQuery } from '@tanstack/react-query';
import { fetchBreeds } from '../../repository/fetch-breeds.ts';

export const queryKeyBreeds = 'breeds';
export const useBreeds = () => {
  const { isPending, error, data } = useQuery({
    queryKey: [queryKeyBreeds],
    queryFn: fetchBreeds,
  });

  return {
    data,
    error,
    isPending,
  };
};
