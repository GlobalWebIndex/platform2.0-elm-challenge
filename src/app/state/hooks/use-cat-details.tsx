import { useQuery } from '@tanstack/react-query';
import { fetchCatDetails } from '../../repository/fetch-cat-details.ts';

export const queryKeyCatDetails = 'cat-details';
export const useCatDetails = (id: string) => {
  const { isPending, error, data } = useQuery({
    queryKey: [queryKeyCatDetails, id],
    queryFn: () => fetchCatDetails(id),
  });

  return {
    data,
    error,
    isPending,
  };
};
