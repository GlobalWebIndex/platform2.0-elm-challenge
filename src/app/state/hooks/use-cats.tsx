import { useQuery } from '@tanstack/react-query';
import { fetchCats } from '../../repository/fetch-cats.ts';
import { useCatPaginationContext } from '../cat-pagination-context/cat-pagination-context.tsx';

export const queryKeyCats = 'cats';
export const useCats = () => {
  const { limit } = useCatPaginationContext();

  const { isPending, error, data } = useQuery({
    queryKey: [queryKeyCats, limit],
    queryFn: () => fetchCats(String(limit)),
  });

  return {
    data,
    error,
    isPending,
  };
};
