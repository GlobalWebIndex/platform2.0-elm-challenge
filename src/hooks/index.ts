import { useEffect, useState } from 'react';
import { useLocation } from 'react-router-dom';

export function useData<T>(
  callback: () => Promise<T>,
  handleError?: () => void
) {
  const [data, setData] = useState<T | null>(null);

  useEffect(() => {
    async function fetchData() {
      try {
        const payload = await callback.call(null);

        setData(payload);
      } catch (error) {
        handleError?.();
      }
    }

    if (data === null) fetchData();
    // eslint-disable-next-line
  }, []);

  // Need this specific code structure
  // to achieve the desired typing
  if (data === null) {
    return [false, null] as const;
  }

  return [true, data] as const;
}

export function usePageTitle() {
  const location = useLocation();
  const title = location.pathname.substring(1);
  const defaultTitle = 'My Cat App';

  return title === '' || title.includes('images') ? defaultTitle : title;
}
