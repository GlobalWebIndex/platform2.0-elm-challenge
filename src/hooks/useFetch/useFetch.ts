import {useCallback, useEffect, useMemo, useState} from 'react';

export const useFetch = <T>(
  endpoint: string
): {abort: () => void; data: T | null; error: Error | null; isFetching: boolean} => {
  const [data, setData] = useState<T | null>(null);
  const [isFetching, setIsFetching] = useState(true);
  const [error, setError] = useState<Error | null>(null);

  const abortController = useMemo(() => new AbortController(), []);

  useEffect(() => {
    fetch(`https://api.thecatapi.com/v1/${endpoint}`, {
      signal: abortController.signal
    })
      .then(response => {
        if (!response.ok) {
          throw new Error('Failed to fetch data');
        }

        return response.json() as T;
      })
      .then(response => {
        setData(response);
      })
      .catch((e: unknown) => {
        if ((e as Error).name === 'AbortError') {
          return;
        }

        setError(e as Error);
      })
      .finally(() => {
        setIsFetching(false);
      });

    return (): void => {
      abortController.abort();
    };
  }, [abortController, endpoint]);

  const abort = useCallback<() => void>(() => {
    abortController.abort();
  }, [abortController]);

  return {abort, data, error, isFetching};
};
