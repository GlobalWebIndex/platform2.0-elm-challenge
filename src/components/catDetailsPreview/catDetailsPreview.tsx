import React from 'react';
import {useParams} from 'react-router';
import type {Cat} from 'cat-lover/types/cats';
import {useFetch} from '../../hooks/useFetch';
import {CatDetails} from '../catDetails';
import {Loader} from '../loader';

export const CatDetailsPreview = (): JSX.Element => {
  const {id} = useParams<'id'>();

  const {data, isFetching} = useFetch<Cat>(`images/${id as string}`);

  if (isFetching) {
    return <Loader />;
  }

  return <CatDetails cat={data as Cat} />;
};
