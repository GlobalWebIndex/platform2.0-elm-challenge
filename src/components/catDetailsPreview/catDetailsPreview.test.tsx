import React from 'react';
import {MemoryRouter} from 'react-router-dom';
import {render} from '@testing-library/react';
import type {Cat} from 'cat-lover/types/cats';
import * as UseFetch from '../../hooks/useFetch';
import {CatDetailsPreview} from './catDetailsPreview';

type UseFetch = typeof UseFetch.useFetch<Cat>;
type UseFetchModule = Omit<typeof UseFetch, 'useFetch'> & {useFetch: UseFetch};

const mockUseFetch = jest.fn<ReturnType<UseFetch>, Parameters<UseFetch>>();
jest.mock<UseFetchModule>('../../hooks/useFetch', () => ({
  useFetch: (...args) => mockUseFetch(...args)
}));

describe('<CatDetailsPreview />', () => {
  describe('when data is fetching', () => {
    beforeEach(() => {
      mockUseFetch.mockReturnValue({
        abort: jest.fn(),
        data: null,
        error: null,
        isFetching: true
      });
    });

    it('should render loader', () => {
      const {container} = render(
        <MemoryRouter initialEntries={['/cats/1']}>
          <CatDetailsPreview />
        </MemoryRouter>
      );

      expect(container).toMatchSnapshot();
    });
  });

  describe('when data is fetched', () => {
    const mockCat: Cat = {
      breeds: [],
      height: 500,
      id: '1',
      url: 'http://example.com/cat.jpg',
      width: 400
    };

    beforeEach(() => {
      mockUseFetch.mockReturnValue({
        abort: jest.fn(),
        data: mockCat,
        error: null,
        isFetching: false
      });
    });

    it('should render cat details', () => {
      const {container} = render(
        <MemoryRouter initialEntries={['/cats/1']}>
          <CatDetailsPreview />
        </MemoryRouter>
      );

      expect(container).toMatchSnapshot();
    });
  });
});
