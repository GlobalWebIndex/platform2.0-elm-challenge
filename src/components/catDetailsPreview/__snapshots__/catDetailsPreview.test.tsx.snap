// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CatDetailsPreview /> when data is fetched should render cat details 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="horizontal withNoBreeds"
    >
      <div
        class="image"
      >
        <img
          alt="Cat"
          src="http://example.com/cat.jpg"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<CatDetailsPreview /> when data is fetching should render loader 1`] = `
<div>
  <div
    class="loader"
  >
    <svg>
      <use
        href="function SvgIcons(props) {
  return React.createElement(
    'svg', 
    Object.assign({}, props, {'data-file-name': SvgIcons.name})
  );
}#icon-cw"
      />
    </svg>
  </div>
</div>
`;
